### PR TITLE
netcdf: conan v2

### PR DESCRIPTION
**Pull request by paulharris**
Finally, a relatively simple recipe.

BUT, I removed an old version of netcdf, as it required an older version of hdf5 in order to build without errors.  The older hdf5 has been dropped from the hdf5 recipe.

Note that this bug was fixed in the newer version already in the recipe,
and there is a newer upstream version that we could add as well.
So I don't see the point of keeping the old buggy version.

### CODE REVIEW

**Comment by toge** [comment]
@paulharris 
Please delete .swp files in PR.

**Comment by paulharris** [comment]
Sorry about that, I didn't notice them go through.

Can I suggest you add to .gitignore these:
```
.*.sw?
.sw?
Session.vim
*~
.undodir
```

**Comment by uilianries** [comment]
> Sorry about that, I didn't notice them go through.
> 
> Can I suggest you add to .gitignore these:
> 
> ```
> .*.sw?
> .sw?
> Session.vim
> *~
> .undodir
> ```

Yes, it would be nice!

**Comment by prince-chrismc** [comment]
https://github.com/conan-io/conan-center-index/pull/13864/files

File: recipes/netcdf/all/conanfile.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
-from conans import CMake, ConanFile, tools
+from conan import ConanFile
+from conan.tools.cmake import CMake, CMakeDeps, CMakeToolchain, cmake_layout
+from conan.tools.files import apply_conandata_patches, export_conandata_patches, get, copy, rm, rmdir, replace_in_file
```
Comment (prince-chrismc) [inline recipes/netcdf/all/conanfile.py:3]: Linter 

```suggestion
from conan.tools.files import apply_conandata_patches, export_conandata_patches, get, copy, rm, rmdir
```

File: recipes/netcdf/all/conanfile.py [hunk 2]
```diff
@@ -32,100 +34,92 @@ class NetcdfConan(ConanFile):
         "netcdf4": True,
         "with_hdf5": True,
         "cdf5": True,
-        "dap": True,
+        "dap": False,
```
Comment (prince-chrismc) [inline recipes/netcdf/all/conanfile.py:37]: Why was this changed? https://github.com/Unidata/netcdf-c/blob/0f29b454d282ea2f894648d533fa1ba045ef40c4/CMakeLists.txt#L513

File: recipes/netcdf/all/conanfile.py [hunk 3]
```diff
@@ -32,100 +34,92 @@ class NetcdfConan(ConanFile):
...
-                # 4.7.4 was built and tested with hdf5/1.12.0
-                # It would be nice to upgrade to 1.12.1,
-                # but when the byterange feature is enabled,
-                # it triggers a compile error that was later patched in 4.8.x
-                # So we will require the older hdf5 to keep the older behaviour.
-                self.requires("hdf5/1.12.0")
-            else:
-                self.requires("hdf5/1.12.1")
-
+                self.requires("hdf5/1.13.1")
```
Comment (prince-chrismc) [inline recipes/netcdf/all/conanfile.py:72]: Linter

```suggestion
            self.requires("hdf5/1.13.1")
```

**Review by prince-chrismc — COMMENTED** [review]
One question and some linter comments

**Comment by paulharris** [comment]
> Please, keep 4.7.4: https://github.com/conan-io/conan-center-index/blob/master/docs/reviewing.md#removing-old-versions

grumble groan i had problems in the past with 4.7.4 due to its dependency on an old HDF5 version that I couldn't get to build - a lot of cmake issues.
I'll give it a try...

File: recipes/netcdf/all/conanfile.py (hunk 2 above)
Comment (paulharris) [reply recipes/netcdf/all/conanfile.py:37]: Note in one of my previous commits: `Set default for 'dap' option to False, to avoid openssl dependency`
I can reenable it if openssl can build on CCI now...

**Comment by paulharris** [comment]
Yay new conan makes 4.7.4 possible :)
Ok so the older version is back in.

Note that will likely have to merge #13605 first ... maybe?

**Comment by paulharris** [comment]
This currently doesn't build as hdf5 needs to merge for a dependency update,
and,
on ProofOfConan, the Windows mingw doesn't build due to a compile problem with libcurl.
https://github.com/eirikb/proof-of-conan/actions/runs/3383752907/jobs/5619952742
Otherwise, I expect it to be fine now.

File: recipes/netcdf/all/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/netcdf/all/conanfile.py]: If openssl doesn't  build that's a bigger problem  😱 

I think it should be enabled, I just saw the new version was merged in too

**Comment by prince-chrismc** [comment]
Just https://github.com/conan-io/conan-center-index/pull/13781#discussion_r1013268301

And waiting for downstream PRs

File: recipes/netcdf/all/conanfile.py (hunk 2 above)
Comment (paulharris) [reply recipes/netcdf/all/conanfile.py]: This is done now

**Comment by paulharris** [comment]
> Just [#13781 (comment)](https://github.com/conan-io/conan-center-index/pull/13781#discussion_r1013268301)

This is done

**Comment by paulharris** [comment]
Not stale, just waiting for HDF5 PR

**Comment by paulharris** [comment]
The day has come!  Green on netcdf!
Let there be reviews!

File: recipes/netcdf/all/conandata.yml [hunk 4]
```diff
@@ -7,12 +7,10 @@ sources:
...
 patches:
   "4.7.4":
-    - base_path: "source_subfolder"
-      patch_file: "patches/4.7.4-0001-cmake-subproject.patch"
-    - base_path: "source_subfolder"
-      patch_file: "patches/4.7.4-0002-cmake-msvc.patch"
-    - base_path: "source_subfolder"
-      patch_file: "patches/4.7.4-0003-cmake-HDF5-cmake-components-wrong.patch"
+    - patch_file: "patches/4.7.4-0001-fix-cmake.patch"
+      patch_description: "fixes for cmake target_link_libraries and other problems"
```
Comment (prince-chrismc) [inline recipes/netcdf/all/conandata.yml:11]:
```suggestion
      patch_description: "fixes for cmake target_link_libraries and using deps"
```

File: recipes/netcdf/all/conandata.yml [hunk 5]
```diff
@@ -7,12 +7,10 @@ sources:
...
+      patch_type: "conan"
   "4.8.1":
-    - base_path: "source_subfolder"
-      patch_file: "patches/4.8.1-0001-fix-cmake.patch"
+    - patch_file: "patches/4.8.1-0001-fix-cmake.patch"
+      patch_description: "fixes for cmake target_link_libraries and other problems"
```
Comment (prince-chrismc) [inline recipes/netcdf/all/conandata.yml:15]:
```suggestion
      patch_description: "fixes for cmake target_link_libraries and using deps"
```

**Review by prince-chrismc — COMMENTED** [review]
Looks super good, just the patch descript was a little vague if you could please apply my suggestion this will be a green tick

**Comment by paulharris** [comment]
Thanks!  Yes those patch descriptions were a bit vague

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]
